src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs

```diff
@@ -600,7 +600,7 @@ private void AddInputsFromDependencyGraphSpec(PackageRestoreInputs packageRestor
             var entryPointProjects = dgFileOutput
                 .Projects
                 .Where(project => dgFileOutput.Restore.Contains(project.RestoreMetadata.ProjectUniqueName, StringComparer.Ordinal))
-                .ToList();
+                .ToHashSet();
```

Comment:
You can achieve that through a dictionary of path to project. We shouldn't hash the PackageSpec. 

HashSet/Dictionary and significantly heavier than lists.

Reply:
Perhaps I jumped the gun with changing this line.

Instead of changing `ToList` to `ToHashSet`, would the following change be correct?

```diff
@@ -616,10 +616,9 @@ private void AddInputsFromDependencyGraphSpec(PackageRestoreInputs packageRestor

             // Filter down to just the requested projects in the file
             // that support transitive references.
-            var v3RestoreProjects = dgFileOutput.Projects
-                .Where(project => (project.RestoreMetadata.ProjectStyle == ProjectStyle.PackageReference
-                    || project.RestoreMetadata.ProjectStyle == ProjectStyle.ProjectJson)
-                    && entryPointProjects.Contains(project));
+            var v3RestoreProjects = entryPointProjects
+                .Where(project => project.RestoreMetadata.ProjectStyle == ProjectStyle.PackageReference
+                                  || project.RestoreMetadata.ProjectStyle == ProjectStyle.ProjectJson);

             packageRestoreInputs.RestoreV3Context.Inputs.AddRange(v3RestoreProjects
                 .Select(project => project.RestoreMetadata.ProjectPath));
```

Since `entryPointProjects` and `v3RestoreProjects` are both filtered versions of `dgFileOutput.Projects`, can we remove the call to `entryPointProjects.Contains` by filtering `entryPointProjects` instead of `dgFileOutput.Projects`?

---

If the above change is not correct, let me expand a bit on why I changed `ToList` to `ToHashSet`.

What I was seeing is that, in a restore for OrchardCore (as an example), `AddInputsFromDependencyGraphSpec` allocates 12.57 MB when using `ToList`, of which `PackageSpec.Equals` is responsible for 12.19 MB.

![image](https://user-images.githubusercontent.com/6613722/173478538-d0dc51c3-d006-4ea8-a796-dea7f076b037.png)

When using `ToHashSet`, `AddInputsFromDependencyGraphSpec` allocates just 1.83 MB.

![image](https://user-images.githubusercontent.com/6613722/173478614-a36efc52-90f8-4c5d-85d2-e22cce0b1b23.png)

A couple of runs of dotTrace shows that with `ToList`, `AddInputsFromDependencyGraphSpec` takes 50-57ms, while using `ToHashSet`, it takes ~40ms, so I'm not worried that we're swapping time for space here.

That said, I'm happy to revert this change and discuss as a separate issue.

Reply:
@nkolev92 
Could we merge as it's or better to revert it back?

Reply:
I'll do a full review and let you know.

Reply:
> Since entryPointProjects and v3RestoreProjects are both filtered versions of dgFileOutput.Projects, can we remove the call to entryPointProjects.Contains by filtering entryPointProjects instead of dgFileOutput.Projects?

That's probably better. 

Given that this is in nuget.exe only, I'm not as concerned about allocations personally. 
We're already doing equality on the full package spec, which allocates less than hashcode but is slower in large solutions.